### PR TITLE
Refactor Tag Manager setup and settings components with datastore

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

Addresses issue #1386, #1608 

Storybook:
https://google.github.io/site-kit-wp/storybook/pull/1715/?path=/story/tag-manager-module-setup--start

## Relevant technical choices

* Added `ContainerSelect` component to handle much of the common parts between web and AMP selects
* Added `isPrimaryAMP` and `isSecondaryAMP` selectors to `core/site`
* Extracted application of existing tag IDs to custom `useExistingTagEffect` hook
* Added a new `createLegacySettingsWrapper` to `stories/utils` (see #1602 which can be updated to refactor the other usage)
* Opted to pass `registry` to `render` for unit tests and in Storybook for simplicity and to allow for separation of common and test-specific setup
* Used component-level decorators in storybook for `beforeEach`-like set up, which worked really nicely
* Added a new `freezeFetch` test utility which is a fetch mock wrapper (similar to `muteFetch`) but for requests that shouldn't return a response
* Tag Manager now only requires readonly scopes (#1608), and has the necessary `autoSubmit` functionality in place, like Analytics 
    - The `Scopes` class already covers the implicit fulfillment of GTM readonly scopes if the user already granted the edit scope 

Note no `trackEvent` calls were present in the old component. (Perhaps we should add some? @marrrmarrr)

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: stories/modules-tagmanager-settings.stories.js [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+	accountID: '',
+	containerID: '',
+	internalContainerID: '',
+	ampContainerID: '',
+	internalAMPContainerID: '',
+	useSnippet: true,
+};
+
+storiesOf( 'Tag Manager Module/Settings', module )
+	.addDecorator( ( storyFn ) => {
```
Comment (felixarntz) [inline stories/modules-tagmanager-settings.stories.js:47]: Nice! So this function runs initially and then calls the actual story function? May be worth refactoring other stories too, although fairly low priority.

File: stories/modules-tagmanager-setup.stories.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+storiesOf( 'Tag Manager Module/Setup', module )
+	.addDecorator( ( storyFn ) => {
+		const registry = createTestRegistry();
+		registry.dispatch( STORE_NAME ).setSettings( {} );
+		registry.dispatch( STORE_NAME ).receiveGetExistingTag( null );
+		registry.dispatch( CORE_USER ).receiveGetAuthentication( {} );
+
+		return storyFn( registry );
+	} )
```
Comment (felixarntz) [inline stories/modules-tagmanager-setup.stories.js:66]: Same for setup, we could at some point update other Storybook files to use a decorator and expanded `Setup` component like here.

File: tests/e2e/plugins/module-setup-tagmanager-no-account.php [hunk 3]
```diff
@@ -20,14 +20,11 @@ function () {
 
 		register_rest_route(
 			REST_Routes::REST_ROOT,
-			'modules/tagmanager/data/accounts-containers',
+			'modules/tagmanager/data/accounts',
```
Comment (felixarntz) [inline tests/e2e/plugins/module-setup-tagmanager-no-account.php:23]: Why was this changed here? Related though: I believe we should remove the `accounts-containers` datapoint entirely now since it's no longer used?

File: assets/js/modules/tagmanager/components/common/AccountCreateLegacy.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+				<Button
+					onClick={ createAccountHandler }
+					target="_blank"
+				>
+					{ __( 'Create an account', 'google-site-kit' ) }
+				</Button>
+
+				<div className="googlesitekit-setup-module__sub-action">
+					<Link onClick={ refetchAccountsHandler }>
+						{ __( 'Re-fetch My Account', 'google-site-kit' ) }
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/AccountCreateLegacy.js:78]: Yeah, I had a similar thought, but yeah, this is how it was.

File: assets/js/modules/tagmanager/components/common/AccountSelect.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+			outlined
+		>
+			{ ( accounts || [] )
+				.concat( {
+					accountId: ACCOUNT_CREATE,
+					name: __( 'Set up a new account', 'google-site-kit' ),
+				} )
+				.map( ( { accountId, name }, index ) => (
+					<Option
+						key={ index }
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/AccountSelect.js:69]: Yeah we can - I didn't realize that using index was discouraged - we do this in most of our selects I believe. I'll update it though.

File: assets/js/modules/tagmanager/components/common/useExistingTagEffect.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
+/**
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/useExistingTagEffect.js:1]: Sure - I put it here because that's how Gutenberg organizes them.
https://github.com/WordPress/gutenberg/tree/master/packages/data/src/components

File: assets/js/modules/tagmanager/components/common/ContainerSelect.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+			{ ...props }
+		>
+			{ ( containers || [] )
+				.concat( {
+					publicId: CONTAINER_CREATE,
+					name: __( 'Set up a new container', 'google-site-kit' ),
+				} )
+				.map( ( { publicId, name, containerId }, index ) => ( // Capitalization rule exception: publicId, containerId
+					<Option
+						key={ index }
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/ContainerSelect.js:69]: Updated to use `publicId` as React complains with `containerId` since it isn't present for the `CONTAINER_CREATE` option.

File: assets/js/modules/tagmanager/datastore/__factories__/builders.js [hunk 8]
```diff
@@ -93,7 +93,21 @@ export const containerBuilder = build( 'Tag Manager Container', {
...
+ * Generates multiple containers.
+ *
+ * @since n.e.x.t
+ *
+ * @param {number} count       Number of containers to generate.
+ * @param {Object} [overrides] Optional. Object of container field overrides.
+ * @return {Object[]} Array of generated container objects.
+ */
+export const buildContainers = ( count, overrides ) => {
+	return Array( count ).fill( undefined )
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/datastore/__factories__/builders.js:105]: Oh that's nice, I didn't know you could do that 👍

File: assets/js/modules/tagmanager/components/common/AccountCreateLegacy.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+			<p>
+				{ __( 'To create a new account, click the button below which will open the Google Tag Manager account creation screen in a new window.', 'google-site-kit' ) }
+			</p>
+			<p>
+				{ __( 'Once completed, click the link below to re-fetch your accounts to continue.', 'google-site-kit' ) }
+			</p>
+
+			<div className="googlesitekit-setup-module__action">
+				<Button
+					onClick={ createAccountHandler }
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/AccountCreateLegacy.js:70]: That's originally how I had it, but it caused problems with the E2E test which hijacks `window.open` to trigger a request to fake that an account has been created. There doesn't seem to be a good way to intercept a link clicked which opens in a new tab in Puppeteer either so I kept it as-is for now.

File: assets/js/modules/tagmanager/components/common/AccountCreateLegacy.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+	const refetchAccountsHandler = useCallback( () => {
+		resetAccounts();
+	} );
+
+	const createAccountHandler = useCallback( () => {
+		// Need to use window.open for this to allow for stubbing in E2E.
+		global.window.open( escapeURI`https://tagmanager.google.com/?authuser=${ userEmail }#/admin/accounts/create`, '_blank' );
+	}, [ userEmail ] );
+
+	if ( undefined === accounts || isDoingGetAccounts ) {
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/AccountCreateLegacy.js:53]: Good call 👍

File: assets/js/modules/tagmanager/datastore/__fixtures__/index.js [hunk 11]
```diff
@@ -16,6 +16,21 @@
...
 
+/**
+ * Internal dependencies
+ */
+import ContainersAMPOnly from './get-containers--amp';
+import ContainersWebOnly from './get-containers--web';
+
 export { default as accounts } from './accounts';
 export { default as createContainer } from './create-container';
 export { default as liveContainerVersion } from './live-container-version.json';
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/datastore/__fixtures__/index.js:27]: @tofumatt mentioned something similar. It works with/without just like `.js` files and I don't think we've ever made a decision about this for json files. I agree they should at least all be consistent so that was a miss here for sure. Not sure if it adds any benefit including `.json` or not though. I'll update them to use `json` extensions.

[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/datastore/__fixtures__/index.js:27]: Updated! Now that it's changed I can see that VS Code offers inspection into the contents of the file where before it didn't, so this is better 👍

File: assets/js/modules/tagmanager/datastore/accounts.js [hunk 12]
```diff
@@ -61,18 +71,50 @@ export const baseActions = {
...
+		invariant( isValidAccountSelection( accountID ), 'A valid accountID selection is required to select.' );
 
-		const registry = yield Data.commonActions.getRegistry();
-		registry.dispatch( STORE_NAME ).setAccountID( accountID );
+		const { select, dispatch } = yield Data.commonActions.getRegistry();
 
-		if ( ACCOUNT_CREATE === accountID ) {
-			// eslint-disable-next-line no-useless-return
+		// Do nothing if the accountID to select is the same as the current.
+		if ( accountID === select( STORE_NAME ).getAccountID() ) {
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/datastore/accounts.js:88]: Sounds good - it has to be in the conditional below though instead since we still want the initial selections to happen.

File: assets/js/modules/tagmanager/datastore/accounts.js [hunk 13]
```diff
@@ -61,18 +71,50 @@ export const baseActions = {
...
+
+		if ( ! isValidAccountID( accountID ) || select( STORE_NAME ).hasExistingTag() ) {
+			return;
+		}
+
+		yield containerActions.waitForContainers( accountID );
 		// Trigger cascading selections.
+		const { isAMP, isSecondaryAMP } = select( CORE_SITE );
+		if ( ! isAMP() || isSecondaryAMP() ) {
+			const webContainers = select( STORE_NAME ).getWebContainers( accountID ) || [];
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/datastore/accounts.js]: Like you said, it shouldn't be possible due to `waitForContainers` above so both of these context-specific selectors should always return an array at this point (the fallback arrays here should actually be removed). I think we should be able to trust that things work as expected - if we can't then we're probably lacking test coverage to give us that. I would prefer adding test coverage than adding additional conditionals for cases that shouldn't be possible.

See the added test here: https://github.com/google/site-kit-wp/pull/1715/commits/c13d44b026f5987b7128bd8254411bd2875c9605

File: package.json [hunk 14]
```diff
@@ -157,7 +157,8 @@
     "@storybook/addon-viewport": "^5.3.18",
     "@storybook/react": "^5.3.18",
     "@testing-library/jest-dom": "^4.1.2",
-    "@testing-library/react": "^9.4.1",
+    "@testing-library/react": "^9.5.0",
+    "@testing-library/react-hooks": "^3.3.0",
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply package.json:161]: `@testing-library/react` was updated to fix a problem with `getByRole` (and perhaps other queries) not supporting a text match on the older version. It's a minor bump from a package I think we can trust to adhere to proper semantic versioning so it should be safe but otherwise I wouldn't have updated it here other than to fix what seemed like a bug.

`@testing-library/react-hooks` is just an addition which is necessary for testing the added hook here. Since it's only used for that it shouldn't affect any other tests.

File: stories/modules-tagmanager-settings.stories.js (hunk 1 above)
Comment (aaemnnosttv) [reply stories/modules-tagmanager-settings.stories.js:47]: Yeah, decorators are intended for "wrapping" your stories. They can be defined at various levels, but this will run for each story in this `storiesOf` call. I'm not sure if it's intended to be used in this way to pass a value to the story function but it works great 😄 .

File: tests/e2e/plugins/module-setup-tagmanager-no-account.php (hunk 3 above)
Comment (aaemnnosttv) [reply tests/e2e/plugins/module-setup-tagmanager-no-account.php:23]: We don't call the datapoint anymore so this was changed for the E2E test to continue to work. Alternatively we could add it as an additional endpoint to mock, but again the former is going to be removed and isn't currently used.

File: assets/js/modules/tagmanager/datastore/containers.js [hunk 15]
```diff
@@ -102,10 +106,83 @@ const baseActions = {
...
+	 * Supports selecting a container that has not been received yet.
+	 *
+	 * @since n.e.x.t
+	 * @private
+	 *
+	 * @param {string} accountID   Tag Manager account ID of container to select.
+	 * @param {string} containerID Tag Manager container `publicId` of container to select.
+	 */
+	*selectContainer( accountID, containerID ) {
+		invariant( isValidAccountID( accountID ), 'A valid accountID is required to select a container.' );
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/datastore/containers.js:122]: Yeah that sounds good 👍

File: assets/js/modules/tagmanager/datastore/containers.js [hunk 16]
```diff
@@ -102,10 +106,83 @@ const baseActions = {
...
+		invariant( isValidContainerID( containerID ), 'A valid containerID is required to select a container.' );
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/datastore/containers.js:123]: Yeah, good call 👍

File: assets/js/modules/tagmanager/components/common/ExistingTagNotice.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+import { sprintf, __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { STORE_NAME } from '../../datastore/constants';
+const { useSelect } = Data;
+
+export default function ExistingTagNotice() {
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/ExistingTagNotice.js:31]: I've combined them for now by moving `ExistingTagNotice` inside `FormInstructions` but kept the component for consistency with the other modules as it may still be useful by itself at some point.

File: assets/js/modules/tagmanager/components/common/FormInstructions.js [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+import Data from 'googlesitekit-data';
+import { STORE_NAME as CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
+import { STORE_NAME } from '../../datastore/constants';
+const { useSelect } = Data;
+
+export default function FormInstructions() {
+	const hasExistingTag = useSelect( ( select ) => select( STORE_NAME ).hasExistingTag() );
+	const isSecondaryAMP = useSelect( ( select ) => select( CORE_SITE ).isSecondaryAMP() );
+
+	if ( hasExistingTag ) {
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/FormInstructions.js:37]: It's only really relevant here since `ExistingTagNotice` returns null if `! hasExistingTag` which includes `undefined`.

I took a look at changing this and I think it's preferable to keep it like this because returning `null` early will cause a flicker every time due to the latency of the tag check request. This would cause the instructions to appear out of nowhere and shift the form down all of a sudden. So I think it's preferable for this flicker to only happen when the message changes due to the presence of an existing tag (which should be much less frequent than when one is present) rather than showing nothing at all until the tag fetch is resolved.

I considered adding our `PreviewBlock` to hold the place as well since some message will always appear, but it was not a great experience so I opted to keep it like this for now. Not saying this is the best way to do it, but I think it's the best way within the scope of this issue.

File: stories/modules-tagmanager-components.stories.js [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+				message: 'They taste like burning!',
+			} );
+		};
+
+		return (
+			<WithTestRegistry callback={ setupRegistry }>
+				<ErrorNotice />
+			</WithTestRegistry>
+		);
+	} )
```
Comment (felixarntz) [inline stories/modules-tagmanager-components.stories.js:135]: I think we can omit this component from here, we don't test it elsewhere and it's really just duplicated across all modules right now. I'll open an issue for implementing a single unified variant of it.

File: stories/modules-tagmanager-components.stories.js [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+import WebContainerSelect from '../assets/js/modules/tagmanager/components/common/WebContainerSelect';
+import AMPContainerSelect from '../assets/js/modules/tagmanager/components/common/AMPContainerSelect';
+import UseSnippetSwitch from '../assets/js/modules/tagmanager/components/common/UseSnippetSwitch';
+import ExistingTagNotice from '../assets/js/modules/tagmanager/components/common/ExistingTagNotice';
+import ExistingTagError from '../assets/js/modules/tagmanager/components/common/ExistingTagError';
+import ErrorNotice from '../assets/js/modules/tagmanager/components/common/ErrorNotice';
+import AccountCreate from '../assets/js/modules/tagmanager/components/common/AccountCreate';
+
+storiesOf( 'Tag Manager Module/Components', module )
+	.add( 'AccountSelect', () => {
```
Comment (felixarntz) [inline stories/modules-tagmanager-components.stories.js:41]: Let's wrap all these components in some setup `div`s like we do in e.g. Analytics - it provides a more representative preview (mainly just because of the white background).

File: stories/modules-tagmanager-components.stories.js [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+				<AccountSelect />
+			</WithTestRegistry>
+		);
+	} )
+	.add( 'WebContainerSelect', () => {
+		const setupRegistry = ( registry ) => {
+			const accountID = fixtures.getContainers.all[ 0 ].accountId;
+			registry.dispatch( STORE_NAME ).receiveGetSettings( {} );
+			registry.dispatch( STORE_NAME ).receiveGetAccounts( [] );
+			registry.dispatch( STORE_NAME ).setAccountID( accountID );
```
Comment (felixarntz) [inline stories/modules-tagmanager-components.stories.js:59]: Nit-picking, but let's use `receiveGetSettings` to define settings that would come from the API rather than using individual calls which are more intended for user interaction.

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.js [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import ContainerSelect from './ContainerSelect';
+import { STORE_NAME } from '../../datastore/constants';
+import { STORE_NAME as CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
```
Comment (tofumatt) [inline assets/js/modules/tagmanager/components/common/AMPContainerSelect.js:30]: Sometimes I wonder about moving these all to a single constants file (or re-exporting them from one), so you could `import { CORE_SITE, CORE_USER, MODULE_ANALYTICS } from 'googlesitekit-data` or something. Just thinking aloud and not related to this issue, but something to ponder 🙂

File: assets/js/modules/tagmanager/components/common/AccountCreate.js [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+				{ __( 'To create a new account, click the button below which will open the Google Tag Manager account creation screen in a new window.', 'google-site-kit' ) }
+			</p>
+			<p>
+				{ __( 'Once completed, click the link below to re-fetch your accounts to continue.', 'google-site-kit' ) }
+			</p>
+
+			<div className="googlesitekit-setup-module__action">
+				<Button
+					onClick={ createAccountHandler }
+					target="_blank"
```
Comment (tofumatt) [inline assets/js/modules/tagmanager/components/common/AccountCreate.js:71]: Does this require `target="_blank"` given it's not a link? The `onClick` handler covers this.

That said: why _isn't_ this just a link rather than a button with an onClick? There doesn't seem to be a benefit to making it a semantic button.

File: assets/js/modules/tagmanager/datastore/__fixtures__/index.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/modules/tagmanager/datastore/__fixtures__/index.js:27]: For what it's worth: as a general rule, I like using filename extensions for anything that _isn't_ code/JS. I suppose it's a bit arbitrary, but I find it helps clarify what a file is—no extension implies JS code, anything else you can infer from the filename.

Nice that it has a bonus of improving VSCode inspection 🙂

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.js (hunk 22 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/AMPContainerSelect.js:30]: Yeah, I've been thinking about that too.

We could do something like this? I would prefer to use this in `registerAllStoresOn` too.
```js
export const STORE_NAME = 'core/site';
export { STORE_NAME as CORE_SITE }
```

File: assets/js/modules/tagmanager/components/common/AccountCreate.js (hunk 23 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/AccountCreate.js:71]: Felix asked about this too, and that's originally how I had it. It's not a great reason but the E2E tests are written to override `window.open` to call our API to fake that an account was created when this is clicked. Since there isn't a great way to intercept a link that opens in a new tab this seems like an acceptable compromise for now. In the future maybe we won't be opening all external links in a new tab 😉 

You're right though that the `target` is useless here and is just leftover from when it was an anchor.

Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/AccountCreate.js]: Fixed 👍

File: assets/js/modules/tagmanager/components/common/AMPContainerSelect.js (hunk 22 above)
Comment (felixarntz) [reply assets/js/modules/tagmanager/components/common/AMPContainerSelect.js:30]: Hmm, I prefer to keep these constants separated since the store names are part of the respective stores, not so much the registry `googlesitekit-data`. _Maybe_ we can create some kind of `default-stores.js` file or so where all of these constants are imported from their actual files and then exported with these more specific names - but I also think having to write the individual imports like now is not really that bad?

Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/components/common/AMPContainerSelect.js:30]: > I also think having to write the individual imports like now is not really that bad?

It's not terrible, but now that we've had to write them 1000 times it would be really nice if we could leverage auto-imports for them, and we can't do that as it is because the exports all have the same name. This would also help ensure consistency in the name that's imported though as well.

The pain for me at least isn't really the alias, it's just looking-up/finding the import which you need to do manually more/less (yes you can copy-paste from another file, but that's also somewhat manual 😉 ).
